pycytominer_transform/convert.py

```diff
@@ -403,46 +277,367 @@ def write_parquet(
-        str reference to read source files from.
-        Note: may be local or remote object-storage location using convention "s3://..." or similar.
-      dest_path: str:
-        Path to write files to.
-        Note: this may only be a local path.
-      source_datatype: Optional[str]: (Default value = None)
-        Source datatype to focus on during conversion.
-      compartments: Optional[List[str]]: (Default value = None)
-        Compartment names to use for conversion.
-      concat: bool:  (Default value = True)
-        Whether to concatenate similar files together.
-      infer_common_schema: bool:  (Default value = True)
-        Whether to infer a common schema when concatenating records.
+        source_path: str:
+            str reference to read source files from.
+            Note: may be local or remote object-storage
+            location using convention "s3://..." or similar.
+        dest_path: str:
+            Path to write files to.
+            Note: this may only be a local path.
+        source_datatype: Optional[str]: (Default value = None)
+            Source datatype to focus on during conversion.
+        compartments: Union[List[str], Tuple[str, ...]]: (Default value = None)
+            Compartment names to use for conversion.
+        metadata: Union[List[str], Tuple[str, ...]]:
+            Metadata names to use for conversion.
+        identifying_columns: Union[List[str], Tuple[str, ...]]:
+            Column names which are used as ID's and as a result need to be
+            ignored with regards to renaming.
+        concat: bool:
+            Whether to concatenate similar files together.
+        join: bool:
+            Whether to join the compartment data together into one dataset
+        joins: str:
+            DuckDB-compatible SQL which will be used to perform the join operations.
+        chunk_columns: Optional[Union[List[str], Tuple[str, ...]]],
+            Column names which appear in all compartments to use when performing join
+        chunk_size: Optional[int],
+            Size of join chunks which is used to limit data size during join ops
+        infer_common_schema: bool:  (Default value = True)
+            Whether to infer a common schema when concatenating records.
+        drop_null: bool:
+            Whether to drop null results.
 
     Returns:
-      Dict[str, List[Dict[str, Any]]]:
-        Grouped records which include metadata about destination filepath
-        where parquet file was written.
+        Union[Dict[str, List[Dict[str, Any]]], str]:
+            Grouped records which include metadata about destination filepath
+            where parquet file was written or a string filepath for the joined
+            result.
     """
 
     # gather records to be processed
     records = gather_records(
         source_path=source_path,
         source_datatype=source_datatype,
-        compartments=compartments,
+        targets=list(compartments) + list(metadata),
```

Comment:
Thank you for asking about this - order shouldn't matter here for source data gathering, but I do think it matters for readability and alignment with other areas (where metadata is acted upon or treated first). I'll change this around here and make similar changes to other areas where this may have occurred.

Reply:
Order for this and related areas has been updated to include metadata first before compartments.